src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -34,12 +34,29 @@ public PipeCompletionCallbacks TryComplete(Exception exception = null)
                 _isCompleted = true;
                 if (exception != null)
                 {
-                    _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+                    // We really want to capture the preceeding stacktrace into the exception,
+                    // if it hasn't already been thrown (and thus isn't a transfer exception),
+                    // but there currently isn't a way of doing that.
+                    CaptureException(ExceptionDispatchInfo.Capture(exception));
                 }
             }
+
             return GetCallbacks();
         }
 
+        private void CaptureException(ExceptionDispatchInfo edi)
+        {
+            try
+            {
+                // We throw and catch to fuse the captured stack into the exception; as it may be used multiple times.
+                edi.Throw();
+            }
+            catch (Exception ex)
+            {
+                _exception = ex;
```

Comment:
Looks like need to throw+catch to get an exception with the Edi capture fused into it; when can then use as the inner exception